File: app/src/main/java/com/chooloo/www/callmanager/service/CallManager.java
package com.chooloo.www.callmanager.service;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.telecom.Call;
import android.telecom.VideoProfile;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.chooloo.www.callmanager.entity.Contact;
import com.chooloo.www.callmanager.ui.base.BaseActivity;
import com.chooloo.www.callmanager.util.ContactUtils;
import com.chooloo.www.callmanager.util.PermissionUtils;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;

import timber.log.Timber;

public class CallManager {

    public static Call sCall;

    public static void answer() {
        if (sCall != null) {
            sCall.answer(VideoProfile.STATE_AUDIO_ONLY);
        }
    }

    public static void reject() {
        if (sCall != null) {
            if (sCall.getState() == Call.STATE_RINGING) {
                sCall.reject(false, null);
            } else {
                sCall.disconnect();
            }
        }
    }

    public static void hold(boolean isHold) {
        if (sCall != null) {
            if (isHold) {
                sCall.hold();
            } else {
                sCall.unhold();
            }
        }
    }

    public static void keypad(char c) {
        if (sCall != null) {
            sCall.playDtmfTone(c);
            sCall.stopDtmfTone();
        }
    }

    public static void registerCallback(Call.Callback callback) {
        if (sCall != null) {
            sCall.registerCallback(callback);
        }
    }

    public static void unregisterCallback(Call.Callback callback) {
        if (sCall != null) {
            sCall.unregisterCallback(callback);
        }
    }

    public static int getState() {
        if (sCall != null) {
            return sCall.getState();
        }
        return Call.STATE_DISCONNECTED;
    }

    public static void call(@NonNull BaseActivity activity, @NonNull String number) {
        if (PermissionUtils.isDefaultDialer(activity)) {
            Intent callIntent = new Intent(Intent.ACTION_CALL, Uri.parse("tel:" + Uri.encode(number)));
            activity.startActivity(callIntent);
        } else {
            PermissionUtils.ensureDefaultDialer(activity);
            activity.showError("Set Koler as your default dialer to make calls");
        }
    }

    public static void callVoicemail(@NonNull Context context) {
        try {
            Intent intent = new Intent(Intent.ACTION_CALL, Uri.parse("voicemail:1"));
            context.startActivity(intent);
        } catch (SecurityException e) {
            Toast.makeText(context, "Couldn't start Voicemail", Toast.LENGTH_LONG).show();
            Timber.e(e);
        }
    }

    public static Contact getContact(Context context) {
        try {
            String number = URLDecoder.decode(sCall.getDetails().getHandle().toString(), "utf-8").replace("tel:", "");
            if (number.contains("voicemail")) {
                return Contact.VOICEMAIL;
            }
            return ContactUtils.lookupContact(context, number);
        } catch (UnsupportedEncodingException e) {
            return Contact.UNKNOWN;
        }
    }
}

File: app/src/main/java/com/chooloo/www/koler/service/CallManager.kt
package com.chooloo.www.koler.service

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.telecom.Call
import android.telecom.VideoProfile
import android.widget.Toast
import com.chooloo.www.koler.entity.Contact
import com.chooloo.www.koler.ui.base.BaseActivity
import com.chooloo.www.koler.util.isDefaultDialer
import com.chooloo.www.koler.util.lookupContact
import com.chooloo.www.koler.util.requestDefaultDialer
import timber.log.Timber
import java.net.URLDecoder

object CallManager {
    @JvmField var sCall: Call? = null
    @JvmField val state: Int = sCall?.state ?: Call.STATE_CONNECTING

    @JvmStatic
    fun answer() {
        sCall?.answer(VideoProfile.STATE_AUDIO_ONLY)
    }

    @JvmStatic
    fun reject() {
        if (sCall?.state == Call.STATE_RINGING) {
            sCall?.reject(false, null)
        } else {
            sCall?.disconnect()
        }
    }

    @JvmStatic
    fun hold(isHold: Boolean) {
        if (isHold) sCall?.hold() else sCall?.unhold()
    }

    @JvmStatic
    fun keypad(c: Char) {
        sCall?.playDtmfTone(c)
        sCall?.stopDtmfTone()
    }

    @JvmStatic
    fun registerCallback(callback: Call.Callback?) {
        sCall?.registerCallback(callback)
    }

    @JvmStatic
    fun unregisterCallback(callback: Call.Callback?) {
        sCall?.unregisterCallback(callback)
    }

    @JvmStatic
    fun call(activity: BaseActivity, number: String?) {
        if (activity.isDefaultDialer()) {
            val callIntent = Intent(Intent.ACTION_CALL, Uri.parse("tel:" + Uri.encode(number)))
            activity.startActivity(callIntent)
        } else {
            activity.requestDefaultDialer()
            activity.showError("Set Koler as your default dialer to make calls")
        }
    }

    @JvmStatic
    fun callVoicemail(context: Context) {
        try {
            val intent = Intent(Intent.ACTION_CALL, Uri.parse("voicemail:1"))
            context.startActivity(intent)
        } catch (e: SecurityException) {
            Toast.makeText(context, "Couldn't start Voicemail", Toast.LENGTH_LONG).show()
            Timber.e(e)
        }
    }

    @JvmStatic
    fun getContact(context: Context): Contact {
        return try {
            val number = URLDecoder.decode(sCall?.details?.handle.toString(), "utf-8").replace("tel:", "")
            if (number.contains("voicemail")) {
                Contact.VOICEMAIL
            } else {
                lookupContact(context, number)
            }
        } catch (e: Exception) {
            Contact.UNKNOWN
        }
    }
}